src/utils/s3-sync.ts

```diff
@@ -0,0 +1,145 @@
+import {
+    DeleteObjectsOutput,
+    DeleteObjectsRequest,
+    ListObjectsV2Output,
+    ListObjectsV2Request,
+    PutObjectOutput,
+    PutObjectRequest,
+    Object as S3Object,
+} from "aws-sdk/clients/s3";
+import * as fs from "fs";
+import * as util from "util";
+import * as path from "path";
+import * as crypto from "crypto";
+import { lookup } from "mime-types";
+import { chunk } from "lodash";
+import chalk from "chalk";
+import { Provider } from "../types/serverless";
+
+const readdir = util.promisify(fs.readdir);
+const stat = util.promisify(fs.stat);
+
+type S3Objects = Record<string, S3Object>;
+
+/**
+ * Synchronize a local folder to a S3 bucket.
+ *
+ * @return True if some changes were uploaded.
+ */
+export async function s3Sync(aws: Provider, localPath: string, bucketName: string): Promise<boolean> {
+    let hasChanges = false;
+    const filesToUpload: string[] = await listFilesRecursively(localPath);
+    const existingS3Objects = await s3ListAll(aws, bucketName);
+
+    // Upload files 5 by 5
+    for (const batch of chunk(filesToUpload, 2)) {
+        await Promise.all(
+            batch.map(async (file) => {
+                const fileContent = fs.readFileSync(path.join(localPath, file));
+
+                // Check that the file isn't already uploaded
+                if (file in existingS3Objects) {
+                    const existingObject = existingS3Objects[file];
+                    const etag = computeS3ETag(fileContent);
+                    if (etag === existingObject.ETag) {
+                        console.log(chalk.gray(`Skipping ${file}, no changes`));
+
+                        return;
+                    }
+                }
+
+                console.log(`Uploading ${file}`);
+                await s3Put(aws, bucketName, file, fileContent);
+                hasChanges = true;
+            })
+        );
+    }
+
+    const objectsToDelete = findObjectsToDelete(Object.keys(existingS3Objects), filesToUpload);
+    if (objectsToDelete.length > 0) {
+        objectsToDelete.map((key) => console.log(`Deleting ${key}`));
+        await s3Delete(aws, bucketName, objectsToDelete);
+        hasChanges = true;
+    }
+
+    return hasChanges;
+}
+
+async function listFilesRecursively(directory: string): Promise<string[]> {
+    const items = await readdir(directory);
+
+    const files = await Promise.all(
+        items.map(async (fileName) => {
+            const fullPath = path.join(directory, fileName);
+            const fileStat = await stat(fullPath);
+            if (fileStat.isFile()) {
+                return [fileName];
+            } else if (fileStat.isDirectory()) {
+                const subFiles = await listFilesRecursively(fullPath);
+
+                return subFiles.map((subFileName) => path.join(fileName, subFileName));
+            }
+
+            return [];
+        })
+    );
+
+    return files.flat(1);
```

Comment:
use Set rather than array, you'll be much more efficient to test for the existence of an item in existing files and not in S3 and the other way around

Reply:
I gave it a try but I'm not sure I got what you meant: https://github.com/getlift/lift/compare/s3-sync...s3-sync-set

With a Set we don't gain much, and the code ends up more complex. Is that what you had in mind?

Reply:
You can use `Set.add` api to progressively build your set in the previous `items.map` using `for item of items` instead. You can use `Set.values()` rather than  destructuring where you need to iterate over the set values (or you can `for myValue of mySet`). But if you don't see too much benefit from using it, keep it that way :)

Reply:
I gave it a quick try and I see what you meant by "use `Set.add`": https://github.com/getlift/lift/compare/s3-sync...s3-sync-set

However I must use restructuring inside `chunk()` because chunk takes an array (even `.values()` doesn't work because it doesn't return an array).

Anyway to be honest here the change is 1:1 in terms of LOC and complexity, so I'd rather merge. But that was interesting to learn about that, thanks!